W-17951162-Review-DW-expressions-config-CP (#435)
mulesoft/docs-code-builder: modules/ROOT/pages/int-configure-dw-expressions.adoc

= Configuring DataWeave Expressions
:page-aliases: int-use-dw-to-transform-data.adoc, int-address-dw-errors.adoc, int-import-dw-libraries.adoc

> Configure DataWeave expressions in your integrations and implementations from *Expression Builder*, *Transformation Builder*, or from the configuration XML. DataWeave is the programming language designed by MuleSoft for data transformation and expressions in connector operations and other components that process messages (payload and attributes) and other Mule event data.

Reviewer: ```suggestion
Configure DataWeave expressions in your integrations and implementations from *Expression Builder*, *Transformation Builder*, or from the configuration XML. DataWeave is the programming language designed by MuleSoft for data transformation and expressions in connector operations and other components that process messages, such as payload and attributes, and other Mule event data.
```

Author: OK! Thanks!

> image::int-dw-transform-ui.png["The *Expression Field*, *Expression Builder*, and *Transformation Builder* button location highlighted in a component."]

Reviewer: ```suggestion
image::int-dw-transform-ui.png["The *Expression Field*, *Expression Builder*, and *Transformation Builder* button location highlighted in a component"]
```

Author: OK! Thanks!

For more resources, see xref:dataweave::index.adoc[]. 

[[open-fx-field]]
== Open the Expression Field 

Most connector operations and components provide at least one field that accepts DataWeave expressions and scripts. In the configuration panel for a component, the *fx* button indicates that the field accepts an expression. In the XML, these fields have the `#[]` markup. Expression fields accept DataWeave expressions, including DataWeave functions and Mule event data, such as `payload`, `attributes`, and `vars`. 

To open a component expression field (*fx*): 

. Open a component that accepts expressions from the canvas in your implementation or integration project. 
+
. Click *fx* to change the field from a string field to an expression field.
+

> image::int-dw-fx-field-set-variable.png["Expression (fx) field in the *Set Variable* component."]

Reviewer: ```suggestion
image::int-dw-fx-field-set-variable.png["Expression (fx) field in the *Set Variable* component"]
```

Author: OK! Thanks!

+
The markup `#[]` in the *fx* field also appears in the XML to indicate that the field is for an expression.
+
----
<set-variable variableName="variableName" value="#[]" 
              doc:name="Set variable" doc:id="ndpiap" />
----
. Click the *Expression Builder* button to configure the component.
+

> image::int-dw-fx-expression-builder.png["*Expression Builder* for the `Set Variable` component."]

Reviewer: ```suggestion
image::int-dw-fx-expression-builder.png["*Expression Builder* for the `Set Variable` component"]
```

Author: OK! Thanks!

* To display a list of DataWeave functions from the Core module from an empty *fx* field, press Ctrl+Space in an _empty_ *fx* field. For example:
+

> image::int-dw-fx-field-autocomplete.png["*Expression Builder* for the `Set Variable` component."]

Reviewer: ```suggestion
image::int-dw-fx-field-autocomplete.png["*Expression Builder* for the `Set Variable` component"]
```

Author: OK! Thanks!

image::dataweave-function-preview-output.png["Preview Output tab]

== Open the Transformation Builder

Create DataWeave transformations in your Mule applications with the *Transformation Builder* tool.

> image::int-dw-transform-builder-ui.png["The *Transformation Builder* window and its sections highlighted."]

Reviewer: ```suggestion
image::int-dw-transform-builder-ui.png["The *Transformation Builder* window and its sections highlighted"]
```

Author: OK! Thanks!

. Click on *Transformation Builder*.
. In the *Transformation Workspace* section, drag the source field from the input payload to the target field to map the source data to the target data.
+

> The mapped fields display a green dot and are highlighted to indicate that the mapping, the *Script* section updates with the new mapping, and the *Output Preview*"* section displays the output of the transformation.

Reviewer: ```suggestion
The mapped fields display a green dot and are highlighted to indicate that the mapping is successful. The *Script* section updates with the new mapping, and the *Output Preview*"* section displays the output of the transformation.
```

Author: Great catch! Thanks!

+

> If needed, map more than one source field to a target field and the Transform Builder concatenates them.

Reviewer: ```suggestion
If needed, map multiple source field to a target field, and then the Transform Builder concatenates them.
```

Author: OK! Thanks!

> . Click on the *More Actions* icon to refresh metadata or remove all mappings.

Reviewer: ```suggestion
. Click the *More Actions* icon to refresh metadata or remove all mappings.
```

Author: OK! Thanks!
